### PR TITLE
controller: fix error retry and add more metrics

### PR DESCRIPTION
**Pull request by nolouch**
### What problem does this PR solve?

Issue Number: Close https://github.com/tikv/pd/issues/8217

### What is changed and how does it work?

```commit-message
controller: fix error retry and add more metrics
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

### Release note

```release-note
None.
```

### CODE REVIEW

**Comment by nolouch** [comment]
ptal @glorv

**Review by rleungx — APPROVED** [review]

File: client/resource_group/controller/controller.go [hunk 1]
```diff
@@ -685,6 +681,30 @@ type groupCostController struct {
...
+type groupMetricsCollection struct {
+	successfulRequestDuration         prometheus.Observer
+	failedLimitReserveDuration        prometheus.Observer
+	requestRetryCounter               prometheus.Counter
+	failedRequestCounterWithOthers    prometheus.Counter
+	failedRequestCounterWithThrottled prometheus.Counter
+	tokenRequestCounter               prometheus.Counter
+}
+
+func initMetrics(oldName, name string) *groupMetricsCollection {
```
Comment (HuSharp) [inline client/resource_group/controller/controller.go:693]: When will we remove `oldName`?

File: client/resource_group/controller/controller.go [hunk 2]
```diff
@@ -789,7 +806,7 @@ func (gc *groupCostController) initRunState() {
 	case rmpb.GroupMode_RUMode:
 		gc.run.requestUnitTokens = make(map[rmpb.RequestUnitType]*tokenCounter)
 		for typ := range requestUnitLimitTypeList {
-			limiter := NewLimiterWithCfg(now, cfgFunc(getRUTokenBucketSetting(gc.meta, typ)), gc.lowRUNotifyChan)
+			limiter := NewLimiterWithCfg(now, cfgFunc(getRUTokenBucketSetting(gc.meta, typ)), gc.lowRUNotifyChan).SetName(gc.name).SetupMetrics()
```
Comment (HuSharp) [inline client/resource_group/controller/controller.go:809]: Can we put `setName` logic in `NewLimiterWithCfg`?

File: client/resource_group/controller/controller.go (hunk 1 above)
Comment (nolouch) [reply client/resource_group/controller/controller.go:693]: maybe 8.5

**Review by lhy1024 — COMMENTED** [review]
Does it need to be picked to 7.5? 7.5.2 will frozen code tomorrow

**Review by HuSharp — APPROVED** [review]

File: client/resource_group/controller/limiter.go [hunk 3]
```diff
@@ -106,7 +116,7 @@ func NewLimiter(now time.Time, r Limit, b int64, tokens float64, lowTokensNotify
 
 // NewLimiterWithCfg returns a new Limiter that allows events up to rate r and permits
 // bursts of at most b tokens.
-func NewLimiterWithCfg(now time.Time, cfg tokenBucketReconfigureArgs, lowTokensNotifyChan chan<- struct{}) *Limiter {
+func NewLimiterWithCfg(name string, now time.Time, cfg tokenBucketReconfigureArgs, lowTokensNotifyChan chan<- struct{}) *Limiter {
 	lim := &Limiter{
 		limit:               Limit(cfg.NewRate),
```
Comment (HuSharp) [inline client/resource_group/controller/limiter.go:121]:
```suggestion
	lim := &Limiter{
		limit:               Limit(cfg.NewRate),
		name: 	       name,
```

**Pull request merged**